File: sdk/npm/src/providers/chatgpt-org-uk.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'chatgpt-org-uk';
const BASE_URL = 'https://mail.chatgpt.org.uk/api';
const HOME_URL = 'https://mail.chatgpt.org.uk/';

const DEFAULT_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36',
  'Accept': '*/*',
  'Referer': 'https://mail.chatgpt.org.uk/',
  'Origin': 'https://mail.chatgpt.org.uk',
  'DNT': '1',
};

function extractGmSid(response: Response): string {
  const setCookie = response.headers.get('set-cookie') || '';
  const match = setCookie.match(/gm_sid=([^;]+)/);
  return match ? match[1] : '';
}

async function fetchGmSid(): Promise<string> {
  const response = await fetchWithTimeout(HOME_URL, {
    method: 'GET',
    headers: DEFAULT_HEADERS,
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch gm_sid: ${response.status}`);
  }

  const gmSid = extractGmSid(response);
  if (!gmSid) {
    throw new Error('Failed to extract gm_sid cookie');
  }

  return gmSid;
}

async function fetchGmSidWithRetry(): Promise<string> {
  try {
    return await fetchGmSid();
  } catch (error: any) {
    const message = String(error?.message || error || '').toLowerCase();
    if (message.includes('401') || message.includes('extract gm_sid')) {
      return await fetchGmSid();
    }
    throw error;
  }
}

async function fetchInboxToken(email: string, gmSid: string): Promise<string> {
  const response = await fetchWithTimeout(`${BASE_URL}/inbox-token`, {
    method: 'POST',
    headers: {
      ...DEFAULT_HEADERS,
      'Content-Type': 'application/json',
      'Cookie': `gm_sid=${gmSid}`,
    },
    body: JSON.stringify({ email }),
  });

  if (!response.ok) {
    throw new Error(`Failed to get inbox token: ${response.status}`);
  }

  const data = await response.json();
  const token = data?.auth?.token;
  if (!token) {
    throw new Error('Failed to get inbox token');
  }

  return token;
}

async function fetchInboxTokenWithRetry(email: string): Promise<string> {
  const gmSid = await fetchGmSidWithRetry();
  try {
    return await fetchInboxToken(email, gmSid);
  } catch (error: any) {
    const message = String(error?.message || error || '').toLowerCase();
    if (message.includes('401')) {
      const refreshedGmSid = await fetchGmSidWithRetry();
      return await fetchInboxToken(email, refreshedGmSid);
    }
    throw error;
  }
}

async function fetchEmails(token: string, email: string): Promise<Email[]> {
  if (!token) {
    throw new Error('internal error: token missing for chatgpt-org-uk');
  }
  const encodedEmail = encodeURIComponent(email);
  const response = await fetchWithTimeout(`${BASE_URL}/emails?email=${encodedEmail}`, {
    method: 'GET',
    headers: {
      ...DEFAULT_HEADERS,
      'x-inbox-token': token,
    },
  });

  if (!response.ok) {
    throw new Error(`Failed to get emails: ${response.status}`);
  }

  const data = await response.json();

  if (!data.success) {
    throw new Error('Failed to get emails');
  }

  const rawEmails = data.data?.emails || [];
  return rawEmails.map((raw: any) => normalizeEmail(raw, email));
}

async function fetchEmailsWithRetry(email: string): Promise<Email[]> {
  const token = await fetchInboxTokenWithRetry(email);
  try {
    return await fetchEmails(token, email);
  } catch (error: any) {
    const message = String(error?.message || error || '').toLowerCase();
    if (message.includes('401')) {
      const refreshedToken = await fetchInboxTokenWithRetry(email);
      return await fetchEmails(refreshedToken, email);
    }
    throw error;
  }
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${BASE_URL}/generate-email`, {
    method: 'GET',
    headers: DEFAULT_HEADERS,
  });

  if (!response.ok) {
    throw new Error(`Failed to generate email: ${response.status}`);
  }

  const data = await response.json();

  if (!data.success) {
    throw new Error('Failed to generate email');
  }

  const email = data.data.email;
  const token = await fetchInboxTokenWithRetry(email);

  return {
    channel: CHANNEL,
    email,
    token,
  };
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  if (!token) {
    throw new Error('internal error: token missing for chatgpt-org-uk');
  }

  try {
    return await fetchEmails(token, email);
  } catch (error: any) {
    const message = String(error?.message || error || '').toLowerCase();
    if (message.includes('401')) {
      return await fetchEmailsWithRetry(email);
    }
    throw error;
  }
}

File: sdk/python/tempmail_sdk/providers/chatgpt_org_uk.py
"""
mail.chatgpt.org.uk 渠道实现
API: https://mail.chatgpt.org.uk/api
"""

from urllib.parse import quote
from .. import http as tm_http
from ..types import EmailInfo
from ..normalize import normalize_email

CHANNEL = "chatgpt-org-uk"
BASE_URL = "https://mail.chatgpt.org.uk/api"
HOME_URL = "https://mail.chatgpt.org.uk/"

DEFAULT_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36",
    "Accept": "*/*",
    "Referer": "https://mail.chatgpt.org.uk/",
    "Origin": "https://mail.chatgpt.org.uk",
    "DNT": "1",
}


def _extract_gm_sid(resp) -> str:
    for name, value in resp.cookies.items():
        if name == "gm_sid":
            return value
    return ""


def _fetch_gm_sid_once() -> str:
    resp = tm_http.get(
        HOME_URL,
        headers=DEFAULT_HEADERS,
    )
    resp.raise_for_status()

    gm_sid = _extract_gm_sid(resp)
    if not gm_sid:
        raise Exception("Failed to extract gm_sid cookie")
    return gm_sid


def _fetch_gm_sid() -> str:
    try:
        return _fetch_gm_sid_once()
    except Exception as exc:
        msg = str(exc).lower()
        if "401" in msg or "gm_sid" in msg:
            return _fetch_gm_sid_once()
        raise


def _fetch_inbox_token_once(email: str, gm_sid: str) -> str:
    resp = tm_http.post(
        f"{BASE_URL}/inbox-token",
        headers={
            **DEFAULT_HEADERS,
            "Content-Type": "application/json",
            "Cookie": f"gm_sid={gm_sid}",
        },
        json={"email": email},
    )
    resp.raise_for_status()
    data = resp.json()
    token = data.get("auth", {}).get("token")
    if not token:
        raise Exception("Failed to get inbox token")
    return token


def _fetch_inbox_token(email: str) -> str:
    gm_sid = _fetch_gm_sid()
    try:
        return _fetch_inbox_token_once(email, gm_sid)
    except Exception as exc:
        if "401" in str(exc):
            gm_sid = _fetch_gm_sid()
            return _fetch_inbox_token_once(email, gm_sid)
        raise



def generate_email(**kwargs) -> EmailInfo:
    """创建临时邮箱"""
    resp = tm_http.get(
        f"{BASE_URL}/generate-email",
        headers=DEFAULT_HEADERS,
    )
    resp.raise_for_status()
    data = resp.json()

    if not data.get("success"):
        raise Exception("Failed to generate email")

    email = data["data"]["email"]
    token = _fetch_inbox_token(email)

    return EmailInfo(
        channel=CHANNEL,
        email=email,
        _token=token,
    )



def get_emails(token: str, email: str, **kwargs) -> list:
    """获取邮件列表"""
    if not token:
        raise Exception("internal error: token missing for chatgpt-org-uk")

    def _fetch_emails(token_value: str) -> list:
        resp = tm_http.get(
            f"{BASE_URL}/emails?email={quote(email)}",
            headers={
                **DEFAULT_HEADERS,
                "x-inbox-token": token_value,
            },
        )
        resp.raise_for_status()
        data = resp.json()

        if not data.get("success"):
            raise Exception("Failed to get emails")

        return [normalize_email(raw, email) for raw in (data.get("data", {}).get("emails") or [])]

    try:
        return _fetch_emails(token)
    except Exception as exc:
        if "401" in str(exc):
            refreshed_token = _fetch_inbox_token(email)
            return _fetch_emails(refreshed_token)
        raise
